### ISSUE TITLE
Ability to specify additional compiler flags [imported]

### ISSUE THREAD

**Issue by cmaglie** [arduino/Arduino#421]
This is [Issue 421](http://code.google.com/p/arduino/issues/detail?id=421) moved from a Google Code project.
Added by 2010-11-30T03:12:51.000Z by [steamra...@yahoo.com](http://code.google.com/u/103275024174372104156/).
Please review that bug for more context and additional comments, but update this bug.

Original labels: Type-Enhancement, Priority-Medium, Component-IDE
### Original description

A simple option in preferences.txt to change compiler flags instead of having them hard-wired in would be very nice.

A way to change it in the GUI would be even better.

**Comment by ElectricRCAircraftGuy** [comment]
Hi does anyone know what the compiler settings are currently set to in the Arduino IDE and where can I find this?

**Comment by ElectricRCAircraftGuy** [comment]
I think I found it, though I don't know what I'm looking at exactly. 
"C:\Program Files (x86)\Arduino\hardware\arduino\avr\platform.txt"

**Comment by matthijskooijman** [comment]
Yup, that's the right place. See https://github.com/arduino/Arduino/wiki/Arduino-IDE-1.5---3rd-party-Hardware-specification for (slightly out-of-date) documentation.

**Comment by anders-liu** [comment]
It's really a long-term issue....

But I still want to vote it, because the application-wise definition (-D option) is really important for me.

I'm going to write a library, and I want user to use ```#define MY_FLAG``` to control which will be compiled in my library. I don't want use run time parameter which increase program size and many code never runs in the application lifetime.

If this's hard to be done in IDE, could you add some special 'higher level' comments in code? For example, let user can define compiler options in code (*.ino file) like this:

```
/// [!compiler-options] -D MY_FLAG

#include "library.h"

void setup(){}

void loop() {}
```

You can parse it before send code to compiler, and append/replace the option set in code explicitly.

**Comment by facchinm** [comment]
@anders-liu could something like this work? https://github.com/facchinm/arduino-builder/commit/904250febf1dea20408bdd1d1fa8950195f5b7a1

**Comment by matthijskooijman** [comment]
@facchinm, this looks overpowered to me: AFAICS it allows setting *any* build property, which is really flexible, but also allows overriding e.g. the entire compiler commandline, which allows executing arbitrary commands as well (e.g. `rm -rf /`, or a root shell, or a backdoor installer, etc.). That's not quite what I expect when I compile someone else's code.

I'm not quite sure where the balance lies: Limiting to just #defines might be good (and perhaps make the syntax more specific for that (`#arduino-define FOO=bar`) rather than accepting the full properties and trying to filter that. Perhaps more things make sense than just define. Things like the board to compile for would make sense, though that should probably be handled by the IDE (e.g. preload the board when opening the sketch, but still allow changing it afterwards instead of silently overriding the select board on all builds).

**Comment by facchinm** [comment]
Totally agree, right now that approach is very insecure but I like the cgo approach (see https://golang.org/cmd/cgo/#hdr-Using_cgo_with_the_go_command).
The directives will change from
```
// #cgo CFLAGS: -DPNG_DEBUG=1
// #cgo amd64 386 CFLAGS: -DX86=1
// #cgo LDFLAGS: -lpng
```
into something like
```
// #arduino define: PNG_DEBUG=1
// #arduino arduino:avr define: X86=0
```
which is super readable and debuggable.
Need to discuss about that, anyway :stuck_out_tongue:

**Comment by anders-liu** [comment]
@facchinm First your efforts are great. But why do I need to involve such heavy build process? I just making some toys... If I need the heavy works, why don't I just switch to an IDE, such as Atmel Studio.

**Comment by Himura2la** [comment]
It's very strange that I can not configure libraries even through the build flags. Vote for this issue

**Comment by hank** [comment]
I need to specify the following easily (currently I'm modifying platform.txt directly, which is terrible because it's not portable):

``
compiler.cpp.flags=-c -Os {compiler.warning_flags} -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -MMD -flto -funsigned-char -funsigned-bitfields -fpack-struct -fshort-enums -finline-limit=3 -fno-inline-small-functions -mcall-prologues -Wl,--relax -fno-tree-scev-cprop
``

I like the idea above that would allow me to do something like the following in the sketch:

``
// #arduino compiler.cpp.flags: -c -Os {compiler.warning_flags} -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -MMD -flto -funsigned-char -funsigned-bitfields -fpack-struct -fshort-enums -finline-limit=3 -fno-inline-small-functions -mcall-prologues -Wl,--relax -fno-tree-scev-cprop
``

Making them direct overrides of platform.txt keys would be my preference. That way, it's very obvious what you're modifying. You could also support a simple append functionality if someone just wanted to add a flag (which is probably what I actually need):

``
// #arduino append compiler.cpp.flags: -funsigned-char -funsigned-bitfields -fpack-struct -fshort-enums -finline-limit=3 -fno-inline-small-functions -mcall-prologues -Wl,--relax -fno-tree-scev-cprop
``

Syntax is up for debate; as long as I can do this in the .ino or a .cpp file, I'm pretty comfortable with anything.

**Comment by ArizonaClark** [comment]
FWIW, I have a somewhat different application for the ability to be able to specify `"-D<name>=<value>"` than the uses others have mentioned.

I have multiple very similar hardware units to program that require only minor differences in the source code (which is currently over 1000 lines long), e.g., each has to have a unique IP address, and some have different sensors connected.  I'd very much like to have something like
`#if UNIT == UNIT01`
`#define MY_IP "192.168.0.101"`
`#elif UNIT == UNIT02`
`#define MY_IP "192.168.0.102"`
and so on.  (There are other relatively minor differences besides this simple example.)

Based on a half century of programming, I'd suggest an "Additional Compiler Options" under "Tools" in the IDE.

**Comment by d-a-v** [comment]
This is a very interesting idea @facchinm

> Need to discuss about that, anyway
```
// #arduino define: PNG_DEBUG=1
// #arduino arduino:avr define: X86=0
```
> which is super readable and debuggable.

Would Arduino maintainers agree with such a PRoposal or has anybody already started to be working on such a great feature ?

**Comment by d-a-v** [comment]
I have a proposal for this issue.
Please check arduino/arduino-cli/pull/1117
